Add WebApplication.CreateEmptyBuilder

eerhardt opened the pull request:
Allow an empty WebApplicationBuilder to be created without default behavior. There is no default configuration sources (ex. environment variables or appsettings.json files), no logging, and no web server (ex. Kestrel) configured by default. These can all be added explicitly by the app.

The following middleware can be enabled by the app:

- Routing and Endpoints, if the app calls Services.AddRouting, and then MapXXX or registers an EndpointDataSource manually
- AuthN/Z, if the app adds the corresponding Auth services

Fix #48811

cc @DamianEdwards @davidfowl

eerhardt committed:
Add WebApplication.CreateEmptyBuilder

Allow an empty WebApplicationBuilder to be created without default behavior. There is no default configuration sources (ex. environment variables or appsettings.json files), no logging, and no web server (ex. Kestrel) configured by default. These can all be added explicitly by the app.

The following middleware can be enabled by the app:

- Routing and Endpoints, if the app calls MapXXX or registers an EndpointDataSource manually
- AuthN/Z, if the app adds the corresponding Auth services
- HostFiltering and ForwardedHeaders

Fix #48811

davidfowl commented on src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:
```diff
@@ -1666,8 +1761,21 @@ public async Task WebApplicationBuilder_OnlyAddsDefaultServicesOnce(CreateBuilde
+
+        // These services are still necessary
+        Assert.Single(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IWebHostEnvironment)));
+        Assert.Single(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IOptionsChangeTokenSource<HostFilteringOptions>)));
```
Why is this part of the empty builder?

eerhardt replied:
It comes from this code

https://github.com/dotnet/aspnetcore/blob/bd9f5d1df2137935bec66b711c7c1534978a421c/src/DefaultBuilder/src/WebHost.cs#L265-L268

from my understanding, there is no other public way to get host filtering enabled. The class is internal and no other code references it.

eerhardt replied:
I remember @blowdart saying that HostFiltering is recommended. See https://github.com/dotnet/aspnetcore/issues/47862.

davidfowl replied:
We don’t want another empty builder . This has to be the last low level builder API so it really needs to be empty

eerhardt replied:
Fixed.

eerhardt committed:
PR feedback

mitchdenny commented on src/DefaultBuilder/src/WebHost.cs:
```diff
@@ -244,9 +244,15 @@ internal static void ConfigureWebDefaultsSlim(IWebHostBuilder builder)
+        {
+            // Routing services need to be registered in order for WebApplication to
+            // implement IEndpointRouteBuilder (i.e. MapGet, etc) correctly.
+            services.AddRoutingCore();
```
Why even call AddRoutingCore?

DamianEdwards replied:
'cos then `app.MapGet` methods and friends don't work. Routing is a core requirement in the `WebApplication` model.

eerhardt replied:
Also for the simple reasoning of "because @halter73 told me to".

eerhardt replied:
The one drawback I can think of having this is that AddRoutingCore does bring in app size that may not be required.

If I publish the following app for NativeAOT:

```C#
var builder = WebApplication.CreateEmptyBuilder(new ());
builder.WebHost.UseKestrelCore();

var app = builder.Build();

app.Use(async (context, next) =>
{
    await context.Response.WriteAsync("Hello, World!");
    await next(context);
});

app.Run();
```

on win-x64 I get:

* 7.98 MB with AddRoutingCore
* 7.50 MB without

So AddRoutingCore adds roughly 500 KB of app size, even when it isn't used.

DamianEdwards replied:
How does it fail if you try to call `MapGet` without adding the routing services & middleware?

davidfowl replied:
CreateEmptyBuilder doesn’t need to be usable by default. It just needs to be empty. We can make sure the errors are good but empty should really mean empty.

captainsafia replied:
> How does it fail if you try to call `MapGet` without adding the routing services & middleware?

You get the following exception at startup:

```
Unhandled exception. System.InvalidOperationException: Unable to find the required services. Please add all the required services by calling 'IServiceCollection.AddRouting' inside the call to 'ConfigureServices(...)' in the application startup code.
```

If we were to not register the routing services by default, you'd start off with the following code:

```csharp
var builder = WebApplication.CreateEmptyBuilder(new());
var app = builder.Build();
app.MapGet("/", () => "Hello world!");
app.Run();
```

You'd hit an exception about no `IServer` implementation being registered which you'd fix with this:

```diff
var builder = WebApplication.CreateEmptyBuilder(new());
+builder.WebHost.UseKestrel();
var app = builder.Build();
app.MapGet("/", () => "Hello world!");
app.Run();
```

Then you'd get the exception about the lack of routing services which you'd fix with this:

```diff
var builder = WebApplication.CreateEmptyBuilder(new());
builder.WebHost.UseKestrel();
+builder.Services.AddRouting();
var app = builder.Build();
app.MapGet("/", () => "Hello world!");
app.Run();
```

Before you have a functioning app.

> We can make sure the errors are good but empty should really mean empty.

There's a chance we can improve the experience you get here by producing a single exception that outlines the full requirements for WebApplication instead of two.

Although I hear the argument about routing services being required to fulfill the spirit of the `WebApplication` (as in it implement `IEndpointBuilder`), it feels like if we're going through the effort to make a _truly_ empty builder, then we shouldn't compromise on that principal.

captainsafia replied:
One other thing that struck me weird about this is the fact that everywhere else in the framework (AFAIK) we register middleware in response to services existing in DI. This is the first example I've seen where we eagerly register services in response to anticipating we'll add a certain middleware.

mitchdenny replied:
Routing shouldn't be required for this right (ignoring current implementation details for a moment):

```csharp
var builder = WebApplication.CreateEmptyBuilder();
var app = builder.Build();
app.Use(async (context, next) => await next(context)); 
app.Run();
```

mitchdenny replied:
Assume ```Use(...)``` does something interesting and is the end of the line :)

captainsafia replied:
> Routing shouldn't be required for this right (ignoring current implementation details for a moment):

No, you don't need routing services for that case.

halter73 replied:
I was going to suggest adding the AddRoutingCore services conditionally on a EndpointDataSource being added to the WebApplication similar to what we do for the middleware, but then realized the service provider has already been built.

It'd be nice if we could make it "just work", but that doesn't seem easy. I'm okay with leaving out AddRoutingCore as long as the error is clear enough which it appears to be. Maybe we should update the exception message to recommend AddRouting or AddRoutingCore.

mitchdenny replied:
The other thing to keep in mind is that we don't need to really worry about whether this would work out of the box because it's unlikely that we'll ever have this in a template where that is a requirement.

We tend to analyse these decisions as if looking at what the minimal working example would be (generally a good idea). But that isn't a requirement here.

eerhardt replied:
I've removed AddRoutingCore from the empty builder.

I didn't update the exception message because it is used in a few places and tries to be generic. Updating it to include "or AddRoutingCore" would have made it be not generic anymore. So I left it. We can do it in a separate PR if required.

eerhardt committed:
Address PR feedback.

Remove Routing from the "Empty" builder. It is up to the app to configure what they want. This is the Empty WebApplication.

halter73 commented on src/DefaultBuilder/src/WebApplicationBuilder.cs:
```diff
@@ -154,9 +145,76 @@ internal WebApplicationBuilder(WebApplicationOptions options, bool slim, Action<
+
+                webHostBuilder.UseSetting(WebHostDefaults.ApplicationKey, _hostApplicationBuilder.Environment.ApplicationName ?? "");
+
+                // NOTE: There is no way to add Configuration before this gets called, so it is unnecessary
```
This is not true given that `options.Args` can be passed in, and those should be respected. This feels like an important escape hatch in case someone using the empty builder wants to modify the hosting startup assemblies.

eerhardt replied:
Fixed and refactored.

halter73 commented on src/DefaultBuilder/src/WebApplicationBuilder.cs:
```diff
@@ -154,9 +145,76 @@ internal WebApplicationBuilder(WebApplicationOptions options, bool slim, Action<
+        // This is for testing purposes
+        configureDefaults?.Invoke(bootstrapHostBuilder);
+
+        bootstrapHostBuilder.ConfigureSlimWebHost(
```
I forgot about this public API. It is a little weird that the "empty" builder configures the "slim" web host.

As a follow up, would it make more sense to rename this to `ConfigureEmptyWebHost` and then add a `ConfigureSlimWebHost` that also calls `ConfigureWebDefaultsSlim` and adds Kestrel and the core routing services?

eerhardt replied:
I've opened https://github.com/dotnet/aspnetcore/issues/49375 to track this.

eerhardt replied:
Note that these are the things added in ConfigureSlimWebHost:

https://github.com/dotnet/aspnetcore/blob/224026829144defe56e878f4011b7a54ca10353c/src/Hosting/Hosting/src/GenericHost/SlimWebHostBuilder.cs#L20-L60

It doesn't do anything with Kestrel or routing. That is higher up in the stack.

Looking through these services, the only ones I could imagine removing in an "EmptyWebHost" are all related to observability:

* DiagnosticListener
* ActivitySource
* DistributedContextPropagator
* AddMetrics/HostingMetrics

halter73 replied:
> It doesn't do anything with Kestrel or routing. That is higher up in the stack.

I realize it's that way now. I was suggesting moving `ConfigureWebDefaultsSlim` into `ConfigureSlimWebHost` as part of the change.

eerhardt replied:
It can't be because they live in different assemblies - thus why we needed to make that API public.

For example `ConfigureWebDefaultsSlim` uses stuff in `Microsoft.AspNetCore.dll`, like `HostFilteringStartupFilter` and in references like Routing, which `Microsoft.AspNetCore.Hosting.dll` doesn't have.

halter73 commented on src/DefaultBuilder/src/WebApplicationBuilder.cs:
```diff
@@ -154,9 +145,76 @@ internal WebApplicationBuilder(WebApplicationOptions options, bool slim, Action<
+        _genericWebHostServiceDescriptor = InitializeHosting(bootstrapHostBuilder);
+    }
+
+    internal WebApplicationBuilder(WebApplicationOptions options, bool slim, bool empty, Action<IHostBuilder>? configureDefaults = null)
```
Nit: We only have two remaining readonly fields. Is it time to make those non readonly (and possibly "lie" about the nullability), give these more logical method names and remove the weird bool args. E.g. `InitializeEmpty`, `InitializeSlim`, `InitializeFull`?

eerhardt replied:
Let's consider this in a future change. I'm not excited about the factoring now, but it can easily be changed later without breaking anyone.

halter73 reviewed (approved):
I think the only thing that really needs to be addressed before merging is the hosting startup stuff when configured via the command line.

eerhardt committed:
Address PR feedback.

Respect hosting startup configuration. Refactor to share logic.
